File: 1. Study/0704 - 그래프와 탐색(DFS)/kimshin/3. 미로탐색(DFS).py
arr = [list(map(int, input().split())) for _ in range(7)]
cnt = 0
dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]
arr[0][0] = 1

def dfs(x, y):
    global cnt
    if x == 6 and y == 6:
        cnt += 1
        return
    else:
        for i in range(4):
            nx, ny = x + dx[i], y + dy[i]
            if 0 <= nx < 7 and 0 <= ny < 7 and arr[nx][ny] == 0:
                arr[nx][ny] = 1
                dfs(nx, ny)
                arr[nx][ny] = 0

dfs(0, 0)
print(cnt)

File: 1. Study/0704 - 그래프와 탐색(DFS)/mojaeya/3_미로탐색(DFS).js
let arr = [
    [0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 1, 1, 1, 1, 1, 0],
    [0, 0, 0, 0, 1, 0, 0, 0],
    [0, 1, 1, 0, 1, 0, 1, 1],
    [0, 1, 1, 0, 0, 0, 0, 1],
    [0, 1, 1, 0, 1, 1, 0, 0],
    [0, 1, 0, 0, 0, 0, 0, 0]
  ];

console.log(solution(arr));

function solution(board) {
    let cnt = 0;
    let path = [];
    let dx = [-1, 0, 1, 0];
    let dy = [0, 1, 0, -1];

    function DFS(x, y) {
      if(x === 7 && y === 7) {
          console.log(path)
          cnt++;
      }
      else {
        for(let k = 0; k < 4; k++) {
          let nx = x + dx[k];
          let ny = y + dy[k];
          if(nx >= 1 && nx <= 7 && ny >= 1 && ny <= 7 && !board[nx][ny]) { 
            board[nx][ny] = 1;
            path.push(`[${nx}][${ny}]`)
            DFS(nx, ny); 
            board[nx][ny] = 0; 
            path.pop();
          }
        }
      }
    }
    board[1][1] = 1;
    DFS(1, 1);
    return cnt;
  }